components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigration.kt

```diff
@@ -0,0 +1,216 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.support.migration
+
+import android.content.Context
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.async
+import mozilla.components.browser.storage.sync.PlacesBookmarksStorage
+import mozilla.components.browser.storage.sync.PlacesHistoryStorage
+import mozilla.components.support.base.log.logger.Logger
+import mozilla.components.support.base.utils.SharedPreferencesCache
+import org.json.JSONArray
+import org.json.JSONObject
+import java.io.File
+import java.lang.Exception
+import java.lang.IllegalStateException
+import java.util.concurrent.Executors
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * Supported Fennec migrations.
+ */
+enum class Migration {
+    HISTORY, BOOKMARKS, OPEN_TABS
+}
+
+/**
+ * Result of running a set of [Migration]s. Maps a [Migration] to a tuple of (migration version, [MigrationResult]).
+ */
+typealias MigrationsResult = Map<Migration, Pair<Int, MigrationResult>>
+
+// Describes ordering and versioning of migrations.
+private val migrations = listOf(
+    Migration.HISTORY to 1,
+    Migration.BOOKMARKS to 1,
+    Migration.OPEN_TABS to 1
+)
+private data class MigrationRun(val version: Int, val success: Boolean)
+private typealias MigrationHistory = Map<Migration, MigrationRun>
+
+private class MigrationHistoryStore(context: Context) : SharedPreferencesCache<MigrationHistory>(context) {
+    override val logger = Logger("MigrationRecord")
+    override val cacheKey = "MigrationRecord"
+    override val cacheName = cacheKey
+
+    override fun MigrationHistory.toJSON(): JSONObject {
+        return JSONObject().also { result ->
+            result.put("list", JSONArray().also { list ->
+                this.forEach {
+                    list.put(JSONObject().also { run ->
+                        run.put("migration", it.key.name)
+                        run.put("version", it.value.version)
+                        run.put("success", it.value.success)
+                    })
+                }
+            })
```

Comment:
random nit: I don't feel like the fancy use of `also` chains here makes this actually easier to read than just sequential code. :)

Reply:
\/\o/\/ I kind of like it, it looks nested :)